packages/protocol/contracts/Staking.sol

```diff
@@ -0,0 +1,687 @@
+     * @param _indexes Indexes of the locks to be withdrawn.
+     * @return totalAmount_ Total amount of tokens withdrawn.
+     */
+    function withdraw(
+        address _subject,
+        uint256[] memory _indexes
+    ) external returns (uint256 totalAmount_) {
+        address subjectTokenAddress;
+
+        (subjectTokenAddress, totalAmount_) = _extractExpiredAndDeleteLocks(
+            _subject,
+            _indexes
+        );
+
+        IERC20Extended subjectToken = IERC20Extended(subjectTokenAddress);
+
+        subjectToken.safeTransfer(msg.sender, totalAmount_);
+        emit Withdraw(
+            msg.sender,
+            _subject,
+            subjectTokenAddress,
+            _indexes,
+            totalAmount_
+        );
+    }
+
+    /**
+     * External function to extend the lock period of the tokens.
+     * @param _subject Subject address for which tokens are being extended.
+     * @param _indexes Indexes of the locks to be extended.
+     * @param _lockPeriodInSec New lock period for the tokens.
+     * @return totalAmount_ Total amount of tokens extended for lock.
+     * @return unlockTimeInSec_ unlock time is the timestamp of the extended lock.
+     */
+    function extendLock(
+        address _subject,
+        uint256[] memory _indexes,
+        uint256 _lockPeriodInSec
+    )
+        external
+        onlyValidLockPeriod(_lockPeriodInSec)
+        returns (uint256 totalAmount_, uint256 unlockTimeInSec_)
+    {
+        address subjectToken;
+        (subjectToken, totalAmount_) = _extractExpiredAndDeleteLocks(
+            _subject,
+            _indexes
+        );
+        emit LockExtended(
+            msg.sender,
+            _subject,
+            subjectToken,
+            _indexes,
+            totalAmount_
+        );
+        (, uint256 unlockTimeInSec) = _createLock(
+            _subject,
+            totalAmount_,
+            _lockPeriodInSec,
+            msg.sender,
+            false
+        );
+
+        unlockTimeInSec_ = unlockTimeInSec;
+    }
+
+    /**
+     * External function to extend the lock period of the tokens.
+     * @param _subject Subject address for which tokens are being extended.
+     * @param _indexes Indexes of the locks to be extended.
+     * @param _lockPeriodInSec New lock period for the tokens.
+     * @param _beneficiary Address of the beneficiary for the lock.
+     * @return totalAmount_ Total amount of tokens extended for lock.
+     * @return unlockTimeInSec_ unlock time is the timestamp of the extended lock.
+     */
+    function extendLockFor(
+        address _subject,
+        uint256[] memory _indexes,
+        uint256 _lockPeriodInSec,
+        address _beneficiary
+    )
+        external
+        onlyValidLockPeriod(_lockPeriodInSec)
+        returns (uint256 totalAmount_, uint256 unlockTimeInSec_)
+    {
+        address subjectToken;
+        (subjectToken, totalAmount_) = _extractExpiredAndDeleteLocks(
```

Comment:
ideally, this function should accept the beneficiary and check for lock expiry instead of referring msg.sender

Reply:
only lock beneficiary can withdraw or extend. so msg.sender is valid check